### PR TITLE
feat(useWebNotification): add `requestPermissions` option, return `permissionGranted` and `ensurePermissions`

### PR DESCRIPTION
**Pull request by michealroberts**
---

### Description

This MR addresses the usage concerns in #3314

This MR **_does not_** introduce breaking changes to the usage of the functional API, but introduces a change in the triggering of the notification request permission, i.e., it **will no longer** request permission `onMounted`.

---

### <samp>🤖 Generated by Copilot at 093dd7a</samp>

This pull request enhances the `useWebNotification` hook with a new option and helper functions to request permission for web notifications more flexibly and conveniently. It also updates the documentation with an example of using the new option.

### <samp>🤖 Generated by Copilot at 093dd7a</samp>

*  Add `requestOnMounted` option to `useWebNotification` hook to control when permission is requested ([link](https://github.com/vueuse/vueuse/pull/3325/files?diff=unified&w=0#diff-941615d6e0fe6606acafbb3b72343a981ce06a4090861827cfdf1f7b329b8a28L84-R92),[link](https://github.com/vueuse/vueuse/pull/3325/files?diff=unified&w=0#diff-941615d6e0fe6606acafbb3b72343a981ce06a4090861827cfdf1f7b329b8a28L96-R119),[link](https://github.com/vueuse/vueuse/pull/3325/files?diff=unified&w=0#diff-941615d6e0fe6606acafbb3b72343a981ce06a4090861827cfdf1f7b329b8a28L145-R164),[link](https://github.com/vueuse/vueuse/pull/3325/files?diff=unified&w=0#diff-80106f8d8e5faad5e5d95114a4dd39a986bd2fe105733a2c7452f7607ce3ae1dR54-R68))
*  Simplify permission checking and requesting logic using `hasPermission` and `requestPermission` functions ([link](https://github.com/vueuse/vueuse/pull/3325/files?diff=unified&w=0#diff-941615d6e0fe6606acafbb3b72343a981ce06a4090861827cfdf1f7b329b8a28L110-R127),[link](https://github.com/vueuse/vueuse/pull/3325/files?diff=unified&w=0#diff-941615d6e0fe6606acafbb3b72343a981ce06a4090861827cfdf1f7b329b8a28L121-R144),[link](https://github.com/vueuse/vueuse/pull/3325/files?diff=unified&w=0#diff-941615d6e0fe6606acafbb3b72343a981ce06a4090861827cfdf1f7b329b8a28R189-R190))
*  Refactor `defaultOptions` to `defaultWebNotificationOptions` and destructure `window` option ([link](https://github.com/vueuse/vueuse/pull/3325/files?diff=unified&w=0#diff-941615d6e0fe6606acafbb3b72343a981ce06a4090861827cfdf1f7b329b8a28L96-R119),[link](https://github.com/vueuse/vueuse/pull/3325/files?diff=unified&w=0#diff-941615d6e0fe6606acafbb3b72343a981ce06a4090861827cfdf1f7b329b8a28L121-R144))
*  Update `useWebNotification/index.md` with new option and functions ([link](https://github.com/vueuse/vueuse/pull/3325/files?diff=unified&w=0#diff-80106f8d8e5faad5e5d95114a4dd39a986bd2fe105733a2c7452f7607ce3ae1dR54-R68))

### CODE REVIEW

**Comment by antfu** [comment]
Updated the interface to align with https://vueuse.org/core/useDevicesList/#usedeviceslist,

Also set the value default to true to avoid breaking

**Review by antfu — APPROVED** [review]

**Comment by SampsonCrowley** [comment]
@michealroberts I'm a little late to the party but I think there's a typo here

File: packages/core/useWebNotification/index.ts [hunk 1]
```diff
@@ -118,11 +137,13 @@ export function useWebNotification(defaultOptions: UseWebNotificationOptions = {
 
   // Show notification method:
   const show = async (overrides?: WebNotificationOptions) => {
-    if (!isSupported.value)
+    // If either the browser does not support notifications or the user has
+    // not granted permission, do nothing:
+    if (!isSupported.value && !permissionGranted.value)
```
Comment (SampsonCrowley) [inline packages/core/useWebNotification/index.ts:142]: shouldn't this be `||`

`if (!isSupported.value || !permissionGranted.value)`

**Comment by SampsonCrowley** [comment]
https://github.com/vueuse/vueuse/pull/3422

**Pull request merged**